app/queries/reasons_for_rejection_count_query.rb

```diff
@@ -0,0 +1,106 @@
+class ReasonsForRejectionCountQuery
+  THIS_MONTH = 'this_month'.freeze
+  BEFORE_THIS_MONTH = 'before_this_month'.freeze
+
+  Result = Struct.new(:all_time, :this_month, :sub_reasons)
+
+  def reason_counts
+    rows = ActiveRecord::Base.connection.exec_query(
+      reason_counts_sql,
+      'SQL',
+      [[nil, Time.zone.now.beginning_of_month]],
+    ).to_a
+
+    to_results(rows)
+  end
+
+  def sub_reason_counts
+    results = reason_counts
+
+    rows = ActiveRecord::Base.connection.exec_query(
+      sub_reason_counts_sql,
+      'SQL',
+      [[nil, Time.zone.now.beginning_of_month]],
+    ).to_a
+
+    add_sub_results(results, rows)
+  end
+
+private
+
+  MAPPING = {
+    candidate_behaviour_what_did_the_candidate_do: :candidate_behaviour_y_n,
+    quality_of_application_which_parts_needed_improvement: :quality_of_application_y_n,
+    qualifications_which_qualifications: :qualifications_y_n,
+    honesty_and_professionalism_concerns: :honesty_and_professionalism_y_n,
+    safeguarding_concerns: :safeguarding_y_n,
+  }.with_indifferent_access
+
+  def to_results(rows)
+    results_hash = ActiveSupport::HashWithIndifferentAccess.new do |hash, reason|
+      hash[reason] = Result.new(
+        0, 0, ActiveSupport::HashWithIndifferentAccess.new do |sub_hash, sub_reason|
+          sub_hash[sub_reason] = Result.new(0, 0, nil)
+        end
+      )
+    end
+
+    rows.each_with_object(results_hash) do |row, results|
+      if row['time_period'] == THIS_MONTH
+        results[row['key']].this_month += row['count'].to_i
+      end
+      results[row['key']].all_time += row['count'].to_i
+    end
+  end
+
+  def add_sub_results(results, rows)
+    rows.each do |row|
+      result = results[MAPPING[row['key']]]
```

Comment:
Optional: I wonder if it's worth extracting this logic into some explicit methods. It might just be me, but it took me a while to figure out exactly what was going on here. 

Maybe something like 

```
def get_result_object_from_rows_key
end

def get_nested_sub_result_from_rows_value
end 

def populate_sub_result_counts
end 
```

Reply:
that's better, it's not just you!